extension/js/common/message-renderer.ts

```diff
@@ -386,19 +396,35 @@ export class MessageRenderer {
   };
 
   public deleteExpired = (): void => {
-    this.processedMessages.deleteExpired();
+    void this.processedMessages.deleteExpired();
     this.downloader.deleteExpired();
   };
 
   public msgGetProcessed = async (msgId: string): Promise<ProcessedMessage> => {
-    let processed = this.processedMessages.get(msgId);
-    if (!processed) {
-      processed = (async () => {
-        return this.processFull(await this.downloader.msgGetFull(msgId));
-      })();
-      this.processedMessages.set(msgId, processed);
+    // Couldn't use caching mechanism in firefox
+    // https://github.com/FlowCrypt/flowcrypt-browser/pull/5651#issuecomment-2054128442
+    if (Catch.isFirefox()) {
+      return await this.processFull(await this.downloader.msgGetFull(msgId));
     }
-    return this.processedMessages.await(msgId, processed);
+    // Couldn't use async await for chunkDownloads.get
+    // because if we call `await chunkDownloads.get`
+    // then return type becomes Buf|undfined instead of Promise<Buf>|undfined
+    return new Promise((resolve, reject) => {
+      this.processedMessages
+        .get(msgId)
+        .then(async processed => {
+          if (!processed || Object.keys(processed).length < 1) {
+            processed = (async () => {
+              return this.processFull(await this.downloader.msgGetFull(msgId));
+            })();
+          }
+          await this.processedMessages.set(msgId, processed);
```

Comment:
@ioanmo226 I've just noticed that all `ExpirationCache` instances use the same `session` storage. So when we set value for `msgId` here, it overwrites value previously set in `await this.fullMessages.set(msgId, msgDownload);` as they use the same `msgId` key. Is it correct or I'm missing something?

Reply:
https://github.com/FlowCrypt/flowcrypt-browser/pull/5651#discussion_r1593217109
@sosnovsky I thought prefix thing was added and now noticed that I reverted that change including reverting to promise based.
Let me revert this prefix thing.

Reply:
Nice catch :+1: